Add missing `config.DISABLE_PERFORMANCE_WARNINGS` checks

maxkapur opened the pull request:
There are a few places where Numba raises a `NumbaPerformanceWarning` without first checking whether `config.DISABLE_PERFORMANCE_WARNINGS` is enabled, meaning that disabling performance warnings (e.g. using the `NUMBA_DISABLE_PERFORMANCE_WARNINGS` environment variable) fails to silence some warnings.

This PR adds these checks.

Closes https://github.com/numba/numba/issues/9762 (the associated bug).

Closes https://github.com/numba/numba/pull/9767 (PR providing an alternative solution).

maxkapur committed:
Add missing `config.DISABLE_PERFORMANCE_WARNINGS` checks

gmarkall reviewed (dismissed):
This looks good to me. A couple of thoughts:

- The CUDA changes will also need porting over to https://github.com/NVIDIA/numba-cuda where continued development of the CUDA target is taking place - if you're happy to, please feel welcome to open a similar PR there. If you'd prefer not to, let me know and I'll open an issue there to add to to the to-do list.
- There aren't really comprehensive tests of disabling the warnings, so it's in keeping with the existing tests that there aren't new tests added here. (Personally I'm OK with this).
- A release note should be added to clarify that the performance warning disabling is fixed in the cases that are missed (see e.g. https://github.com/numba/numba/pull/9768/files#diff-9c273bb73233af5c0d5f335b163965e399d6ca89de3ced932a34157ee9b762f8 for an example of one, and https://github.com/numba/numba/tree/main/docs/upcoming_changes for an explanation of the format)

maxkapur committed:
Add release note

maxkapur commented:
Thank you for the feedback! I added the release notes. I also went ahead and rebased on `main` (hence the force-push).

Here is a companion PR for numba-cuda: https://github.com/NVIDIA/numba-cuda/pull/70

maxkapur commented:
Hello @gmarkall, wondering if this slipped through the cracks as it still says "waiting for author."

swap357 commented:
Hi @maxkapur, 
Thanks for the PR. I’ve reviewed it and it looks good. 
Could you please merge main onto this branch and resolve the conflicts?

swap357 committed:
Merge remote-tracking branch 'upstream/main' into fix-disable-performance-warnings-simple

swap357 reviewed (approved):
lgtm, thank you @maxkapur
